### PR TITLE
First step towards rustfix compiletest mode

### PR DESCRIPTION
**Pull request by killercup**
This is the first small step towards testing auto-fixable compiler
suggestions using compiletest. Currently, it only checks if next to a
UI test there also happens to a `*.rs.fixed` file, and then uses rustfix
(added as external crate) on the original file, and asserts that it
produces the fixed version.

To show that this works, I've included one such test. I picked this test
case at random (and because it was simple) -- It is not relevant to the
2018 edition. Indeed, in the near future, we want to be able to restrict
rustfix to edition-lints, so this test cast might go away soon.

In case you still think this is somewhat feature-complete, here's a
quick list of things currently missing that I want to add before telling
people they can use this:

- [x] Make this an actual compiletest mode, with `test [fix] …` output
  and everything
- [x] Assert that fixed files still compile
- [x] Assert that fixed files produce no (or a known set of) diagnostics
  output
- [x] Update `update-references.sh` to support rustfix
- [x] Use a published version of rustfix (i.e.: publish a new version
  rustfix that exposes a useful API for this)

### CODE REVIEW

File: src/tools/compiletest/Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ regex = "0.2"
 serde = "1.0"
 serde_json = "1.0"
 serde_derive = "1.0"
+rustfix = { git = "https://github.com/rust-lang-nursery/rustfix" }
```
Comment (Manishearth) [inline src/tools/compiletest/Cargo.toml]: We should publish before landing

File: src/tools/compiletest/src/autofix.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+fn apply_suggestion(
+    file_content: &mut String,
+    suggestion: &Replacement,
+) -> Result<String, Box<Error>> {
+    use std::cmp::max;
+
+    let mut new_content = String::new();
+
+    // Add the lines before the section we want to replace
+    new_content.push_str(&file_content
```
Comment (Manishearth) [inline src/tools/compiletest/src/autofix.rs]: I feel like this API should be within rustfix itself, ideally shared with the rest of the code.

This also allocates once per suggestion.

Perhaps rustfix should have an API where it applies all suggestions in order, with a callback for asking for confirmation and stuff (which we pass as an empty callback here)


The purpose of this is both to test rustfix and the suggestions, so we need to use the same code as rustfix here.

**Review by Manishearth — CHANGES_REQUESTED** [review]

File: src/tools/compiletest/src/autofix.rs (hunk 2 above)
Comment (killercup) [reply src/tools/compiletest/src/autofix.rs]: Yes, absolutely. This is just the simplest code that works.

**Comment by shepmaster** [comment]
Ping from triage @killercup ! Will you have time to address the review feedback soon?

**Comment by killercup** [comment]
I'll have time to update this PR tomorrow (thank ~~god~~ bors for May Day)

**Comment by killercup** [comment]
@alexcrichton asked how this PR and rustfix in general was doing. 

I was only able to make little progress here today. The PR is rebased and I moved the API for fixing suggestions to the rustfix crate (still unpublished, but that shouldn't be a blocker).

Next immediate step is to add compiling of fixed files and asserting previously observed warnings are gone. After that, filtering of edition-critical lints.

**Comment by alexcrichton** [comment]
Awesome thanks for the cc @killercup! 

I'm curious, though, would it perhaps be best to put tests into the rustfix crate itself? I'd imagine that'd allow for much faster iteration on rustfix itself. Or is the purpose here to test suggestions as they are added to rustc, using what's likely to be a relatively stable base in rustfix?

I can try to clean up some of the todo items here tomorrow, I've had some experience in the past adding new rustfix modes and whatnot.

**Comment by killercup** [comment]
Thanks! Ping me if you have any questions.

The latter: rustfix already has integration tests similar to compiletest
(incl. simple clippy support)—this is to make sure we can trust (and don’t
break) the suggestions rustc produces.

Alex Crichton <notifications@github.com> schrieb am Mi. 2. Mai 2018 um
01:38:

**Comment by alexcrichton** [comment]
Ok I've pushed up a commit that makes a dedicated suite for rustfix tests and populated with a set of working tests from `src/test/ui/suggestions`.

I think the last step here is to publish rustfix to crates.io and get it pulled in from there.

@killercup would it be possible to split out the binary/library to avoid having compiletest depend on dependencies like clap in this repo?

**Comment by killercup** [comment]
Thanks so much, @alexcrichton!

@Manishearth said in https://github.com/rust-lang/rust/pull/50084#issuecomment-382898078 he'd rather see this is part of the UI tests, probably to speed the tests up and to see the regular diagnostics at the same time.

**Comment by alexcrichton** [comment]
@Manishearth can you elaborate on the idea for a UI mode? The implementation here with the necessary bells and whistles seems like it's pretty different from UI and somewhat makes sense as a separate mode?

**Comment by killercup** [comment]
I think the "upgrade UI tests" workflow is pretty powerful: Ideally, we'd be able to make one PR soon that adds `.fixed` files for all the UI tests that already compile after running rustfix, and then open a "help wanted" issue with a list of UI tests whose suggestions are broken. With some good instructions this could get people involved just like the "new diagnostics" effort a few years ago.

That said, it's not important _right now:_ We only care about a handful of lints right now.

**Comment by alexcrichton** [comment]
@killercup oh since it's optional to have the *.fixed file in the first place I figured it's a good assertion for "that file should exist" and it also triggers the mode to write out the expected file in the build directory

Unfortunately *.rs.fixed is actually somewhat tricky due to how crate names are calculated in rustc, so it's actually a good bit easier to not deal with it

File: src/test/ui/str-as-char.fixed [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
```
Comment (Manishearth) [inline src/test/ui/str-as-char.fixed]: This file seems to be duplicated? It also exists under `suggestions/`

File: src/test/ui/str-as-char.fixed [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {
+    println!("●●");
+    //~^ ERROR character literal may only contain one codepoint
```
Comment (Manishearth) [inline src/test/ui/str-as-char.fixed]: kinda wish we could filter these out entirely but this seems annoying

(I'm not quite sure why we require `//~^ ERROR` on ui tests anyway given that it's all in the output)

File: src/test/ui/suggestions/str-as-char.rs [hunk 5]
```diff
@@ -8,6 +8,8 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// run-rustfix
```
Comment (Manishearth) [inline src/test/ui/suggestions/str-as-char.rs]: Can we detect this by the existence of a .fixed file?

File: src/tools/compiletest/src/runtest.rs [hunk 6]
```diff
@@ -2568,6 +2570,21 @@ impl<'test> TestCx<'test> {
...
         errors += self.compare_output("stderr", &normalized_stderr, &expected_stderr);
 
+        if self.config.compare_mode.is_some() {
+            // don't test rustfix with nll right now
+        } else if self.props.run_rustfix {
+            // Apply suggestions from rustc to the code itself
+            let unfixed_code = self.load_expected_output_from_path(&self.testpaths.file)
+                .unwrap();
+            let suggestions = get_suggestions_from_json(&proc_res.stderr, &HashSet::new()).unwrap();
+            let fixed_code = apply_suggestions(&unfixed_code, &suggestions);
```
Comment (Manishearth) [inline src/tools/compiletest/src/runtest.rs]: We also need to stabilize the suggestion applicability stuff and use it here. cc @killercup 

(Perhaps with a directive letting you opt in to non-applicable suggestions)

**Review by Manishearth — APPROVED** [review]
r=me

File: src/test/ui/suggestions/str-as-char.rs (hunk 5 above)
Comment (alexcrichton) [reply src/test/ui/suggestions/str-as-char.rs]: Ah I commented to this [above as well](https://github.com/rust-lang/rust/pull/50084#issuecomment-386401606)

**Pull request merged**